File: CODE/leetcodes/src/main/java/com/hit/math/linked_list/_141.java
package com.hit.math.linked_list;

import com.hit.math.utils.ListNode;

/**
 * 141. Linked List Cycle
 * <p>
 * Given a linked list, determine if it has a cycle in it.
 * <p>
 * Follow up:
 * <p>
 * Can you solve it without using extra space?
 */
public class _141 {
    public boolean hasCycle(ListNode head) {
        if (head == null) {
            return false;
        }
        /**
         * Use two pointers, walker and runner
         */
        ListNode walker = head;
        ListNode runner = head;
        while (runner.next != null && runner.next.next != null) {
            /**
             * walker moves step by step,
             * runner moves two steps at time
             */
            walker = walker.next;
            runner = runner.next.next;
            /**
             * If the Linked List has a cycle walker and runner will meet at some point
             */
            if (walker == runner) {
                return true;
            }
        }
        return false;
    }
}

File: codes/go/leetcodes/interview/easy_collection/141.go
package easy_collection

/**
 * 141. Linked List Cycle
 * <p>
 * Given a linked list, determine if it has a cycle in it.
 * <p>
 * Follow up:
 * <p>
 * Can you solve it without using extra space?
 */

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func hasCycle(head *ListNode) bool {
	fast := head
	slow := head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
		if fast == slow {
			return true
		}
	}
	return false
}
